usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateNoOverlappingStopTimeInTripBlock.java

```diff
@@ -0,0 +1,364 @@
+     * @param otherTripId            other {@link Trip} id
+     * @param tripFirstTime          {@link Trip} first time
+     * @param tripLastTime           {@link Trip} last time
+     * @param otherTripFirstTime     other {@link Trip} first time
+     * @param otherTripLastTime      other {@link Trip} last time
+     * @param otherTripFirstStopTime other {@link Trip} first {@link StopTime}
+     * @param otherTripLastStopTime  other {@link Trip} last {@link StopTime}
+     */
+    private void checkOverlappingTripsDifferentServiceIdCalendarNotProvided(final String blockId,
+                                                                            final Trip trip,
+                                                                            final String tripId,
+                                                                            final Trip otherTrip,
+                                                                            final String otherTripId,
+                                                                            final int tripFirstTime,
+                                                                            final int tripLastTime,
+                                                                            final Integer otherTripFirstTime,
+                                                                            final Integer otherTripLastTime,
+                                                                            final StopTime otherTripFirstStopTime,
+                                                                            final StopTime otherTripLastStopTime) {
+        // interpret data from `calendar_dates.txt` to determine if currentTrip and
+        // otherTrip operate on same days
+        final Map<String, CalendarDate> tripCalendarDateCollection =
+                dataRepo.getCalendarDateAll().get(trip.getServiceId());
+        final Map<String, CalendarDate> unvisitedTripCalendarDateCollection =
+                dataRepo.getCalendarDateAll().get(otherTrip.getServiceId());
+        final Set<String> currentTripServiceDates = new HashSet<>();
+        tripCalendarDateCollection.forEach((date, calendarDate) -> {
+            if (calendarDate.getExceptionType() == ExceptionType.ADDED_SERVICE) {
+                currentTripServiceDates.add(date);
+            }
+        });
+        final Set<String> unvisitedTripServiceDates = new HashSet<>();
+        unvisitedTripCalendarDateCollection.forEach((date, calendarDate) -> {
+            if (calendarDate.getExceptionType() == ExceptionType.ADDED_SERVICE) {
+                unvisitedTripServiceDates.add(date);
+            }
+        });
+        // get the list of common dates during which trips with different service_id
+        // operate: this represent the list of dates where there could be potential time
+        // overlap.
+        final Set<String> potentialConflictingDates =
```

Comment:
Copying the set and then iterating over it to remove duplicates via `retainAll()` is a potentially very expensive operation on datasets like the Netherlands that put all service in calendar_dates.txt.

I think you can combine this code with the loop through the `unvisitedTripCalendarDateCollection` to have a more efficient implementation in time and memory. Basically, as you loop through `unvisitedTripCalendarDateCollection`, do `currentTripServiceDates.contains(calendarDate)`. If it's `true`, then add the date to the potentially conflicting date list (`potentialConflictingDates`).

Like:

```java
        final Set<String> potentialConflictingDates = new HashSet<>();
        unvisitedTripCalendarDateCollection.forEach((serviceId, calendarDate) -> {
            if (calendarDate.getExceptionType() == ExceptionType.ADDED_SERVICE) {
                if (currentTripServiceDates.contains(calendarDate)) {
                    potentialConflictingDates.add(calendarDate);
                }
            }
        });
```

Reply:
Actually, reading https://docs.oracle.com/javase/8/docs/api/java/time/LocalDate.html, it says:

>use of identity-sensitive operations (including reference equality (==), identity hash code, or synchronization) on instances of LocalDate may have unpredictable results and should be avoided.

So you'd need to change the above to use the `String` representation of the date instead of `LocalDate`.